### PR TITLE
Updater: Log info message when receiving signal

### LINKED ISSUE #946: Log an INFO-level message when SIGINT received
Hi ;)

I'm recently experimenting with your nice bot. However, I stumbled upon this during my "learning-process":

I've created a bot, it worked normally. I just did `updater.start_polling` like in the examples, but without using `updater.idle()` (I thought first `start_polling` is blocking).  Exiting with CTRL+C worked immediately, but has thrown a `KeyboardInterrupt`.
After a while I found the `idle()` function, but I was actually a bit confused of CTRL+C's behaviour, because I've configured logging with a minimum level of `INFO`. So I've seen nothing when I pressed the keys. The bot actually needs a while to shutdown, and because nothing was logged I assumed the SIGINT signal wasn't received (/ assuming due to internal threading stuff not properly processed), so I pressed always CTRL+C two times causing a SIGTERM which worked immediately.
SIGINT notifications are only visible when enabling `DEBUG`-level logging. I would suggest to add an `INFO` message, that tells something like `Received SIGINT, shutting down...`, so nobody is worried ;)

### PR DESCRIPTION
**Pull request by Makman2**
Closes #946

### CODE REVIEW

**Comment by Makman2** [comment]
Wondering why coverage decreased, haven't introduced any new branches...

**Comment by Eldinnie** [comment]
@Makman2 Sometimes coverage fluctuates because errors do or don't happen. The important part is that the diff is 100% covered. I'll look more detailed after Christmas. Thanks for contributing.

**Comment by robert-cody** [comment]
`def signal_name_by_code(code):
    return next((n for n in dir(signal) if n.startswith("SIG") and "_" not in n and getattr(signal, n) == code))`

Looks better.

**Comment by Makman2** [comment]
Okay that shifts lookup effort to each call instead of once preparing a dict inside the module. But we shouldn't raise an implicit `StopIteration` in case there's no such signal-code^^ Might have side-effects when used in for-loops (in case that happens). At least we should do this imo:
```python
try:
    return next(...)
except StopIteration:
    raise KeyError('Unknown signal number {}'.format(...))
```
May I call it `signal_name_from_number` though? `by` sounds less descriptive for me + python docs don't mention the term "code" for those signal numbers^^

**Comment by robert-cody** [comment]
@Makman2 you're right about StopIteration. You can name your method as you wish, i'm not this repo developer, just proposed less code solution.

**Comment by bomjacob** [comment]
We in the dev team are hesitant about adding new dependencies (even just for tests). Especially since in this case it doesn't seem to be required as log can already be captured using pure pytest. See [Loggging - pytest documentation](https://docs.pytest.org/en/latest/logging.html#logging) for details.

File: telegram/utils/helpers.py [hunk 1]
```diff
@@ -19,13 +19,26 @@
...
 
 try:
     from html import escape as escape_html  # noqa: F401
 except ImportError:
     from cgi import escape as escape_html  # noqa: F401
 
+
+# From https://stackoverflow.com/questions/2549939/get-signal-names-from-numbers-in-python
+_signames = {v: k
+             for k, v in reversed(sorted(vars(signal).items()))
```
Comment (tsnoam) [inline telegram/utils/helpers.py]: nitpicking here. the `reversed` & `sorted` are useless here as you're creating a dict. please remove them.

**Comment by Makman2** [comment]
> We in the dev team are hesitant about adding new dependencies (even just for tests). Especially since in this case it doesn't seem to be required as log can already be captured using pure pytest. See Loggging - pytest documentation for details.
Could you maybe update the code so it uses that style instead of the extra dependency?

Sure no problem :+1:

File: telegram/utils/helpers.py (hunk 1 above)
Comment (Makman2) [reply telegram/utils/helpers.py]: I think the `sorted` was there for signals with different names but same number, so there's a deterministic outcome. However I haven't understood the `reversed` call too^^

Comment (Makman2) [reply telegram/utils/helpers.py]: And kk `reversed` then picks signal names that are sorted at the beginning. On my system I get these difference when I don't use `reversed`:
```
SIGIOT vs SIGABRT
SIGCLD vs SIGCHLD
SIGPOLL vs SIGIO
```

Though not sure if it's that important here^^

**Comment by Makman2** [comment]
About
> We in the dev team are hesitant about adding new dependencies (even just for tests). Especially since in this case it doesn't seem to be required as log can already be captured using pure pytest. See Loggging - pytest documentation for details.
Could you maybe update the code so it uses that style instead of the extra dependency?

I have tried using this feature, however:
- It requires the installation of `pytest-capturelog`, so a package is anyway needed for this
- Equivalent testing code with that package fails, pytest gets terminated too early:
  ```
  tests/test_updater.py FTerminated
  ```

So I believe using `testfixtures` is currently the easiest solution.

**Comment by tsnoam** [comment]
@Makman2 
We prefer using pytest ecosystem if possible (so if it comes down to adding a new dependency the prefernce will be for it).
Can you please commit your code for testing with pytest to your branch and allow edits for maintainers as described here:
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

we'll try to work on it together before using `testfixtures`.

**Comment by Makman2** [comment]
Done :+1:

File: telegram/utils/helpers.py (hunk 1 above)
Comment (tsnoam) [reply telegram/utils/helpers.py]: Check manpage of `signal(7)` all the above are synonyms. If one name is chosen over another it's merely a matter of luck, as dictionary order is not guaranteed.
On my system I had different results on different executions.

Another option is to leave the signal name alone and just print the number.

**Comment by tsnoam** [comment]
The problem with pytest in the test you've sent is that it catches the signal itself / prevents Updater from catching it.
I assume that testfixtures forks and runs the test in a different process, thus ignoring the issue.
I'll see what can be done here.

**Comment by tsnoam** [comment]
Note to self: need to make sure that this PR works properly on windows as the unitest in question is skipped on appveyor.

**Comment by tsnoam** [comment]
@Makman2  - I believe that I fixed the relevant test. It was a matter of race condition (the signal arrived too soon).

@Eldinnie - Can you please (manually) verify on windows that the new code doesn't break anything?

File: tests/test_updater.py [hunk 2]
```diff
@@ -240,21 +240,24 @@ def _send_webhook_msg(self,
...
 
-        with caplog.at_level(logging.INFO):
+        with caplog.atLevel(logging.INFO):
             updater.idle()
 
-        assert caplog.record_tuples == [(
-            'telegram.ext.updater', 'INFO', 'Received signal {} (SIGTERM), '
-            'stopping...'.format(signal.SIGTERM))]
+        rec = caplog.records()[-1]
+        assert rec.msg.startswith('Received signal {}'.format(signal.SIGTERM))
```
Comment (Makman2) [inline tests/test_updater.py]: just curious, why not asserting for the complete message? :)

File: telegram/utils/helpers.py (hunk 1 above)
Comment (Makman2) [reply telegram/utils/helpers.py]: > If one name is chosen over another it's merely a matter of luck, as dictionary order is not guaranteed.

Looks like it's true, I'm just wondering: the order we iterate over should always produce a guaranteed outcome. `vars(signal).items()` has unique mapping from `name -> number`. Using `sorted` we get always the same list with the same order, and then we are filling the dict with those values, while switching `name` and `number`. If there are numbers that are equal, later elements in the list should overwrite the old ones. Although using dict-comprehension, citing the `dict`-docs:
>  If a key occurs more than once, the last value for that key becomes the corresponding value in the new dictionary.

Comment (Makman2) [reply telegram/utils/helpers.py]: > Another option is to leave the signal name alone and just print the number.

I wouldn't do that, signal numbers differ on OSes for the "same" signals.

**Comment by Eldinnie** [comment]
```> python .\examples\echobot2.py
2018-01-08 15:10:03,051 - telegram.ext.updater - INFO - Received signal 2 (SIGINT), stopping...
>
```
Works fine on windows.

File: telegram/utils/helpers.py (hunk 1 above)
Comment (tsnoam) [reply telegram/utils/helpers.py]: ok. not worth the argument, lets keep it as is.

**Comment by tsnoam** [comment]
Build fails due to unrelated problems fixed in a different PR.
Will merge now. 

@Makman2 Thanks for your contribution.

**Comment by Makman2** [comment]
Thanks for helping me out and bearing with me ;D

**Comment by tsnoam** [comment]
@Makman2 All the thanks are for you

**Pull request merged**